tools/serverpod_cli/lib/src/analyzer/entities/entity_analyzer.dart

```diff
@@ -294,6 +294,46 @@ class SerializableEntityAnalyzer {
             .any((e) => e.className == fieldDefinition.type.className);
   }
 
+  static void _resolveObjectAndIdRelations(
+    ClassDefinition classDefinition,
+    SerializableEntityFieldDefinition fieldDefinition,
+    List<SerializableEntityDefinition> entityDefinitions,
+  ) {
+    // order of execution matters here.
+    _resolveObjectRelationReference(
+      classDefinition,
+      fieldDefinition,
+      entityDefinitions,
+    );
+    _resolveIdRelationTable(
+      classDefinition,
+      fieldDefinition,
+      entityDefinitions,
+    );
+  }
+
+  static void _resolveObjectRelationReference(
+    ClassDefinition classDefinition,
+    SerializableEntityFieldDefinition fieldDefinition,
+    List<SerializableEntityDefinition> entityDefinitions,
+  ) {
+    var relation = fieldDefinition.relation;
+    if (relation is! UnresolvedObjectRelationDefinition) return;
+
+    fieldDefinition.relation = ObjectRelationDefinition(
+      fieldName: relation.fieldName,
+    );
+
+    var field = classDefinition.findField(relation.fieldName);
+    if (field == null) return;
+
+    field.relation = UnresolvedForeignRelationDefinition(
+      referenceFieldName: relation.foreignFieldName,
+      onUpdate: relation.onUpdate,
+      onDelete: relation.onDelete,
+    );
+  }
+
   static void _resolveIdRelationTable(
```

Comment:
Suggestion: Rename to match the type that we are resolving, just like the introduced `_resolveObjectRelationReference`. (Reference should maybe be definition here as well?)

```suggestion
  static void _resolveForeignRelationDefinition(
```